pytential/symbolic/pde/reduce_fmms.py

```diff
@@ -0,0 +1,321 @@
+            densities=tuple(simplify_densities(densities)))
+
+
+def _convert_target_poly_to_int_g(poly, orig_int_g, rhs_int_g):
+    """This does the opposite of :func:`_convert_kernel_to_poly`
+    and converts a polynomial back to a target derivative
+    operator. It is applied to a :class:`~pytential.symbolic.primitives.IntG`
+    object and returns a new instance.
+    """
+    to_pymbolic = SympyToPymbolicMapper()
+
+    result = 0
+    for monom, coeff in poly.terms():
+        kernel = orig_int_g.target_kernel
+        for idim, rep in enumerate(monom):
+            for _ in range(rep):
+                kernel = AxisTargetDerivative(idim, kernel)
+        result += orig_int_g.copy(target_kernel=kernel,
+                source_kernels=rhs_int_g.source_kernels,
+                densities=rhs_int_g.densities) * to_pymbolic(coeff)
+
+    return result
+
+
+class CoefficientCollector(CoefficientCollectorBase):
+    """This extends :class:`pymbolic.mapper.coefficient.CoefficientCollector`
+    by extending the targets to be any expression instead of just algebraic
+    leafs
+    """
+    def __init__(self, source_dependent_variables):
+        self.source_dependent_variables = source_dependent_variables
+
+    def __call__(self, expr):
+        if expr in self.source_dependent_variables:
+            return {expr: 1}
+        return super().__call__(expr)
+
+    def map_algebraic_leaf(self, expr):
+        if expr in self.source_dependent_variables:
+            return {expr: 1}
+        else:
+            return {1: expr}
+
+    rec = __call__
+
+
+def _syzygy_module(m, gens):
+    """Takes as input a module of polynomials with domain `sympy.EX` represented
+    as a matrix and returns the syzygy module as a matrix of polynomials in the
+    same domain.
+    """
+    from sympy.polys.orderings import grevlex
+
+    def _convert_to_matrix(module, *gens):
+        result = []
+        for syzygy in module:
+            row = []
+            for dmp in syzygy.data:
+                row.append(sympy.Poly(dmp.to_dict(), *gens,
+                    domain=sympy.EX).as_expr())
+            result.append(row)
+        return sympy.Matrix(result)
+
+    ring = sympy.EX.old_poly_ring(*gens, order=grevlex)
+    column_ideals = [ring.free_module(1).submodule(*m[:, i].tolist(), order=grevlex)
+                for i in range(m.shape[1])]
+    column_syzygy_modules = [ideal.syzygy_module() for ideal in column_ideals]
+
+    if not column_syzygy_modules:
+        raise ValueError
+
+    intersection = column_syzygy_modules[0]
+    for i in range(1, len(column_syzygy_modules)):
+        intersection = intersection.intersect(column_syzygy_modules[i])
+
+    m2 = intersection._groebner_vec()
+    m3 = _convert_to_matrix(m2, *gens)
+    return m3
+
+
+def _factor_left(mat, axis_vars):
+    """Return the left hand side of the factorisation of the matrix"""
+    return _syzygy_module(_syzygy_module(mat, axis_vars).T, axis_vars).T
```

Comment:
Explain why this is being called twice.

Reply:
Done in 4ed556e70d5221e1458aa794e94802688798a33a